--- php55-pecl-sasl-0.2.0/config.m4~	2014-06-11 10:48:45.000000000 +0300
+++ php55-pecl-sasl-0.2.0/config.m4	2014-06-11 10:49:37.519596338 +0300
@@ -30,9 +30,9 @@
      AC_MSG_ERROR(Please specify path to SASL installation)
   fi
 
-  AC_MSG_CHECKING(for SASL library in $SASL_DIR/lib)
+  AC_MSG_CHECKING(for SASL library in $SASL_DIR/lib64)
   for i in $SEARCH_LIBS; do
-    if test -r $SASL_DIR/lib/lib$i.a -o -r $SASL_DIR/lib/lib$i.$SHLIB_SUFFIX_NAME; then
+    if test -r $SASL_DIR/lib64/lib$i.a -o -r $SASL_DIR/lib64/lib$i.$SHLIB_SUFFIX_NAME; then
       SASL_LIB=$i
       AC_MSG_RESULT(found -l$i)
       break
@@ -39,12 +39,12 @@
 
   if test -z "$SASL_LIB"; then
      AC_MSG_RESULT(not found)
-     AC_MSG_ERROR(Failed to locate a compatible SASL library in $SASL_DIR/lib)
+     AC_MSG_ERROR(Failed to locate a compatible SASL library in $SASL_DIR/lib64)
   fi
 
   LIBSYMBOL=sasl_server_init
   old_LIBS=$LIBS
-  LIBS="$LIBS -L$SASL_DIR/lib"
+  LIBS="$LIBS -L$SASL_DIR/lib64"
   AC_CHECK_LIB($SASL_LIB, $LIBSYMBOL, [AC_DEFINE(HAVE_SASL_LIB,1,[ ])],
                [AC_MSG_ERROR(unknown SASL library or library not found)])
   LIBS=$old_LIBS
